### PR TITLE
Cff op specializer

### PR DESCRIPTION
**Pull request by behdad**
Work-in-progress.  So far just does generalization, NOT specialization.  Sending for high-level review and test development.  cc @anthrotype @miguelsousa

### CODE REVIEW

**Comment by behdad** [comment]
Almost there!  Need a few more commits...

**Comment by behdad** [comment]
I think this is complete.  I'll go hook it up to the pen now.

**Comment by behdad** [comment]
After we fix tests and merge this (and add more tests!), I'll go ahead and simplify t2CharStringPen even more.

**Comment by behdad** [comment]
Here are tests during development.  Should be useful for adding tests:
```
./fonttools cffLib.specializer 100 100 rmoveto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto 200 300 hlineto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto 200 300 hlineto 900 hmoveto 100 vmoveto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto 200 300 hlineto 900 1000 hmoveto 100 vmoveto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto 200 300 hlineto 900 hmoveto 100 vmoveto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto 200 hlineto 900 hmoveto 100 vmoveto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto 200 hlineto 900 hmoveto 100 vmoveto 1 2 3 4 hhcurveto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto 200 hlineto 900 hmoveto 100 vmoveto 1 2 3 4 5 6 7 8 hhcurveto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto 200 300 400  hlineto 900 hmoveto 100 vmoveto 1 2 3 4 5 6 7 8 hhcurveto
./fonttools cffLib.specializer 100 hstem hintmask xyz 100 100 rmoveto 200 300 400  hlineto
./fonttools cffLib.specializer  200 300 400  hlineto
./fonttools cffLib.specializer  200 300 400  hlineto 1 2 3 4 rlineto
./fonttools cffLib.specializer  200 300 400  hlineto 1 2 3 4 rlineto 9 8 7 6 5 4 rrcurveto
./fonttools cffLib.specializer 1 2 3 4 hhcurveto
./fonttools cffLib.specializer 5 1 2 3 4 hhcurveto
./fonttools cffLib.specializer 5 1 2 3 4 6 hhcurveto
./fonttools cffLib.specializer 5 1 2 3 4 6 7 8 9 hhcurveto
./fonttools cffLib.specializer 5 1 2 3 4 6 7 8 9 vvcurveto
./fonttools cffLib.specializer 5 1 2 3 4 6 7 8 9 vvcurveto 1 2 3 4 5 6 7 8 rcurveline
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 rcurveline
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 rlinecurve 
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 hvcurveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 9 hvcurveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8  hvcurveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 10 11 12 13 14 15 hvcurveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16 hvcurveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16 17 hvcurveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 10 11 12 13 14 15 16 17 18 hvcurveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 10 11 12 13  hvcurveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 10 11 12 13 14 hvcurveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 10 11 12 13 14 vhcurveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 10 11 12 13 14 vhcurveto 2 moveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 10 11 12 13 14 vhcurveto 2 rmoveto
./fonttools cffLib.specializer 1 2 3 4 5 6 7 8 10 11 12 13 14 vhcurveto 2 rmoveto
./fonttools cffLib.specializer 2 rmoveto
./fonttools cffLib.specializer 2 5 rmoveto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto
./fonttools cffLib.specializer 
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto 0 9 rlineto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto 0 9 rlineto 0 1 2 3 4 5 rrcurveto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto 0 9 rlineto 0 1 2 3 4 5 rrcurveto 6 7 8 9 0 1 rrcurveto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto 0 9 rlineto 0 1 2 3 4 5 rrcurveto 6 7 8 9 0 1 rrcurveto 0 0 moveto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto 0 9 rlineto 0 1 2 3 4 5 rrcurveto 6 7 8 9 0 1 rrcurveto 0 0 rmoveto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto 0 9 rlineto 6 0 rlineto 0 7 rlineto 0 1 2 3 4 5 rrcurveto 6 7 8 9 0 1 rrcurveto 0 0 rmoveto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto 0 9 rlineto 6 0 rlineto 0 7 rlineto 0 1 2 3 4 5 rrcurveto 6 7 8 9 0 1 rrcurveto 12 13 lineto 0 0 rmoveto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto 0 9 rlineto 6 0 rlineto 0 7 rlineto 0 1 2 3 4 5 rrcurveto 6 7 8 9 0 1 rrcurveto 12 13 rlineto 0 0 rmoveto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto 0 9 rlineto 6 0 rlineto 0 7 rlineto 0 1 2 3 4 5 rrcurveto 6 7 8 9 0 1 rrcurveto 12 13 14 15 rlineto 0 0 rmoveto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto 0 9 rlineto 6 0 rlineto 0 7 rlineto 0 1 2 3 4 5 rrcurveto 6 7 8 9 0 1 rrcurveto 0 3 2 1 0 0 rrcurveto 12 13 14 15 rlineto 0 0 rmoveto
./fonttools cffLib.specializer 2 5 rmoveto 7 8 rmoveto 8 0 rlineto -3 0 rlineto 0 9 rlineto 6 0 rlineto 0 7 rlineto 0 1 2 3 4 5 rrcurveto 6 7 8 9 0 1 rrcurveto 0 3 2 1 0 0 rrcurveto 12 13 14 15 rlineto 0 0 rmoveto 0 0 1 2 3 0 rrcurveto 4 0 5 6 0 0 rrcurveto
```

File: Lib/fontTools/cffLib/specializer.py [hunk 1]
```diff
@@ -0,0 +1,543 @@
...
+	# 0. Generalize commands.
+	#    This ensures that they are in our expected simple form, with each line/curve only
+	#    having arguments for one segment, and using the generic form (rlineto/rrcurveto).
+	#    If caller is sure the input is in this form, they can turn off generalization to
+	#    save time.
+	#
+	# 1. Combine successive rmoveto operations.
+	#
+	# 2. Specialize rmoveto/rlineto/rrcurveto operators into horizontal/vertical variants.
+	#    We specialize into some, made-up, varianats as well, which simplifies following
```
Comment (justvanrossum) [inline Lib/fontTools/cffLib/specializer.py]: varianats -> variants

**Comment by miguelsousa** [comment]
@behdad the tests have been fixed. The failures are now due to bugs. I'm going to work on tests for the specializer.

**Comment by behdad** [comment]
> @behdad the tests have been fixed. The failures are now due to bugs. I'm going to work on tests for the specializer.

Fixed those.  Thanks!  This should work for realworld fonts now, but I have a bug discovered that I'm working on fixing...

**Comment by miguelsousa** [comment]
With these last commits I was able to run it successfully on a whole font. I then diff'ed the glyph shapes before and after and there were zero differences. So things seem quite solid already.

**Comment by behdad** [comment]
Thanks.  For the curious, here's for example a case that is broken currently:
```
./fonttools cffLib.specializer 10 0 1 2 0 0  0 0 1 2 0 1 0 1 3 4 0 0 rrcurveto
```

My logic was flawed.  I'm going to simplify it.  Looks like it cannot be done without dynamic-programming really.  I'll give up and accept to produce suboptimal result for unusual corner cases.

**Comment by miguelsousa** [comment]
The font I referred to is the one at [otf-making-benchmark](https://github.com/adobe-type-tools/otf-making-benchmark). I haven't updated the files yet, but wanted to mention that the CFF table made by makeotf is still a bit smaller.
```
+------+--------+------------------+
| Tag  | Font 1 |      Font 2      |
+------+--------+------------------+
| CFF  |  78029 |  +3325 (+4.26%)  |
| GPOS |  66070 | +51974 (+78.67%) |
| GSUB |  15450 | -2914 (-18.86%)  |
| cmap |  13918 |  -722 (-5.19%)   |
| hmtx |   7768 |        0         |
| name |   4425 | -1539 (-34.78%)  |
| GDEF |    962 |  -162 (-16.84%)  |
| OS/2 |     96 |        0         |
| BASE |     70 |        0         |
| head |     54 |        0         |
| hhea |     36 |        0         |
| post |     32 |        0         |
| maxp |      6 |        0         |
+------+--------+------------------+
```

**Comment by anthrotype** [comment]
@miguelsousa with or without subroutines?

**Comment by miguelsousa** [comment]
with

**Comment by anthrotype** [comment]
just curious, what about with --no-subroutinize option?

**Comment by behdad** [comment]
Yeah, let's separate subroutinization performance from this.

**Comment by miguelsousa** [comment]
good question. These are the numbers when both fonts are NOT subroutinized:
```
+------+--------+------------------+
| Tag  | Font 1 |      Font 2      |
+------+--------+------------------+
| CFF  | 162403 |  +1884 (+1.16%)  |
| GPOS |  66070 | +51974 (+78.67%) |
| GSUB |  15450 | -2914 (-18.86%)  |
| cmap |  13918 |  -722 (-5.19%)   |
| hmtx |   7768 |        0         |
| name |   3105 |  -219 (-7.05%)   |
| GDEF |    962 |  -162 (-16.84%)  |
| OS/2 |     96 |        0         |
| BASE |     70 |        0         |
| head |     54 |        0         |
| hhea |     36 |        0         |
| post |     32 |        0         |
| maxp |      6 |        0         |
+------+--------+------------------+
```
The option `-nS` turns off subroutinization for makeotf.

**Comment by anthrotype** [comment]
keep in mind that fontmake by default calls booleanOperations to remove the overlaps, whereas makeotf does not, as far as I know, does it? 
In that case, try also with fontmake --keep-overlaps option.

**Comment by behdad** [comment]
I'd be surprised if makeotf doesn't remove overlaps.  But yeah, would be nice to look deeper.  One way is to take makeotf output, pass it through cffLib.specializer.specializeProgram and see if that becomes larger... I'm guessing that the difference is coming from somewhere else, not specialization anymore.

**Comment by behdad** [comment]
Or maybe there's an optimization opportunity I've completely missed :). But cannot imagine any, given that I've been staring at the spec for many days now :D.

**Comment by miguelsousa** [comment]
makeotf will use the UFO's `glyphs.com.adobe.type.processedGlyphs` layer whose glyphs have the overlaps removed. So it's really comparing apples to apples.

**Comment by miguelsousa** [comment]
> One way is to take makeotf output, pass it through cffLib.specializer.specializeProgram and see if that becomes larger...

I'll try that

**Comment by anthrotype** [comment]
@miguelsousa Sorry I didn't see that layer. That Regular.ufo in the [otf-making-benchmark](https://github.com/adobe-type-tools/otf-making-benchmark) repo, is it a UFO 2 or 3? Because the metainfo.plist it says "2", but I see it has UFO-3 layers like the one you mentioned, a layercontents.plist, a "data" subfolder, etc.

**Comment by behdad** [comment]
Just fixed a bug that might be relevant :).

**Comment by miguelsousa** [comment]
@anthrotype I believe it's UFO3 compliant except that it advertises itself as UFO2. It's made this way to allow it to open in RoboFont v1.x.

**Comment by anthrotype** [comment]
the different choice of the defaultWidthX between makeotf and ufo2ft might also have some influence on the overall CFF size.

**Comment by anthrotype** [comment]
yeah, but that hack makes the UFO3 features unavailable to UFO3 compliant editors..

**Comment by anthrotype** [comment]
I think that for the comparison to be fair that "processedGlyphs" layer should be swapped with the default layer, and both makeotf and fontmake should use the very same outlines as input, without any sort of pre-processing by other tools.

**Comment by behdad** [comment]
Ok, I *think* I fixed the remaining issues I was aware of.  New code is even more elegant.

**Comment by behdad** [comment]
> the different choice of the defaultWidthX between makeotf and ufo2ft might also have some influence on the overall CFF size.

Oh, the fact that ufo2ft always encodes width, even if it's default...  I just filed that a few hours ago:
https://github.com/googlei18n/ufo2ft/issues/131

**Comment by miguelsousa** [comment]
@behdad that last commit is damaging a lot of glyphs 😢

**Comment by behdad** [comment]
> @behdad that last commit is damaging a lot of glyphs :cry:

Can you pick one out?

**Comment by behdad** [comment]
> Can you pick one out?

Eg. just give me the program for the correct version.

**Comment by behdad** [comment]
> Eg. just give me the program for the correct version.

Ok, I reproduced.  Looking into it.

**Comment by behdad** [comment]
Should be all good now.

**Comment by behdad** [comment]
Ok, I think I'm done with this for now.

Note that for CFF2 varfonts, the specializer need not be changed at all.  We just need to pass special vector types into it that carry x or y values for all masters at the same time.  The vector just needs to implement __add__ and __bool__/__nonzero__.

I think we can merge this now (and add more tests of course).

File: Lib/fontTools/cffLib/specializer.py [hunk 2]
```diff
@@ -0,0 +1,499 @@
...
+	assert a in '0r'
+	return a
+
+def specializeCommands(commands,
+		       ignoreErrors=False,
+		       generalizeFirst=True,
+		       preserveTopology=False,
+		       maxstack=48):
+
+	maxstack -= 3 # AFDKO code uses effective 45 maxstack while the spec says 48.
```
Comment (adrientetar) [inline Lib/fontTools/cffLib/specializer.py]: So is 48 a problem? Also you're decreasing the value regardless of what gets passed as parameter...

Comment (behdad) [inline Lib/fontTools/cffLib/specializer.py]: Probably not.  Was just being conservative.  I'll remove that line...

**Pull request merged**